File: todo-app/backend/service/todo_service.go
package service

import (
	"fmt"
	"strings"

	"todo-backend/models"
	"todo-backend/repository"
)

// TodoService handles business logic for todo operations
type TodoService struct {
	repo *repository.TodoRepository
}

// NewTodoService creates a new TodoService
func NewTodoService(repo *repository.TodoRepository) *TodoService {
	return &TodoService{
		repo: repo,
	}
}

// CreateTodo creates a new todo item with validation
func (s *TodoService) CreateTodo(req *models.CreateTodoRequest) (*models.Todo, error) {
	// Validate request
	if err := s.validateCreateRequest(req); err != nil {
		return nil, err
	}

	// Create todo from request
	todo := &models.Todo{
		Title:       strings.TrimSpace(req.Title),
		Description: strings.TrimSpace(req.Description),
		Priority:    req.Priority,
		DueDate:     req.DueDate,
		Completed:   false, // New todos are always incomplete
	}

	// Set default priority if not specified
	if todo.Priority == "" {
		todo.Priority = models.PriorityMedium
	}

	// Create todo in repository
	err := s.repo.Create(todo)
	if err != nil {
		return nil, fmt.Errorf("failed to create todo: %w", err)
	}

	return todo, nil
}

// GetTodoByID retrieves a todo by its ID
func (s *TodoService) GetTodoByID(id string) (*models.Todo, error) {
	if strings.TrimSpace(id) == "" {
		return nil, fmt.Errorf("todo ID cannot be empty")
	}

	todo, err := s.repo.GetByID(id)
	if err != nil {
		return nil, fmt.Errorf("failed to get todo: %w", err)
	}

	return todo, nil
}

// GetAllTodos retrieves all todos with optional filtering
func (s *TodoService) GetAllTodos(filter *models.TodoFilter) ([]*models.Todo, error) {
	// Validate filter if provided
	if filter != nil {
		if err := s.validateFilter(filter); err != nil {
			return nil, err
		}
	}

	todos, err := s.repo.GetAll(filter)
	if err != nil {
		return nil, fmt.Errorf("failed to get todos: %w", err)
	}

	return todos, nil
}

// UpdateTodo updates an existing todo
func (s *TodoService) UpdateTodo(id string, req *models.UpdateTodoRequest) (*models.Todo, error) {
	if strings.TrimSpace(id) == "" {
		return nil, fmt.Errorf("todo ID cannot be empty")
	}

	// Validate update request
	if err := s.validateUpdateRequest(req); err != nil {
		return nil, err
	}

	// Update todo in repository
	todo, err := s.repo.Update(id, req)
	if err != nil {
		return nil, fmt.Errorf("failed to update todo: %w", err)
	}

	return todo, nil
}

// DeleteTodo deletes a todo by its ID
func (s *TodoService) DeleteTodo(id string) error {
	if strings.TrimSpace(id) == "" {
		return fmt.Errorf("todo ID cannot be empty")
	}

	err := s.repo.Delete(id)
	if err != nil {
		return fmt.Errorf("failed to delete todo: %w", err)
	}

	return nil
}

// ToggleTodoComplete toggles the completion status of a todo
func (s *TodoService) ToggleTodoComplete(id string) (*models.Todo, error) {
	if strings.TrimSpace(id) == "" {
		return nil, fmt.Errorf("todo ID cannot be empty")
	}

	todo, err := s.repo.ToggleComplete(id)
	if err != nil {
		return nil, fmt.Errorf("failed to toggle todo completion: %w", err)
	}

	return todo, nil
}

// GetTodoStats returns statistics about todos
func (s *TodoService) GetTodoStats() (*models.TodoStatsResponse, error) {
	stats, err := s.repo.GetStats()
	if err != nil {
		return nil, fmt.Errorf("failed to get todo stats: %w", err)
	}

	return stats, nil
}

// validateCreateRequest validates a create todo request
func (s *TodoService) validateCreateRequest(req *models.CreateTodoRequest) error {
	if req == nil {
		return fmt.Errorf("create request cannot be nil")
	}

	if strings.TrimSpace(req.Title) == "" {
		return fmt.Errorf("title cannot be empty")
	}

	if len(req.Title) > 200 {
		return fmt.Errorf("title cannot exceed 200 characters")
	}

	if len(req.Description) > 1000 {
		return fmt.Errorf("description cannot exceed 1000 characters")
	}

	if req.Priority != "" && !models.IsValidPriority(req.Priority) {
		return fmt.Errorf("invalid priority: %s", req.Priority)
	}

	return nil
}

// validateUpdateRequest validates an update todo request
func (s *TodoService) validateUpdateRequest(req *models.UpdateTodoRequest) error {
	if req == nil {
		return fmt.Errorf("update request cannot be nil")
	}

	// Check if at least one field is being updated
	if req.Title == nil && req.Description == nil && req.Completed == nil && 
	   req.Priority == nil && req.DueDate == nil {
		return fmt.Errorf("at least one field must be provided for update")
	}

	if req.Title != nil {
		if strings.TrimSpace(*req.Title) == "" {
			return fmt.Errorf("title cannot be empty")
		}
		if len(*req.Title) > 200 {
			return fmt.Errorf("title cannot exceed 200 characters")
		}
	}

	if req.Description != nil && len(*req.Description) > 1000 {
		return fmt.Errorf("description cannot exceed 1000 characters")
	}

	if req.Priority != nil && !models.IsValidPriority(*req.Priority) {
		return fmt.Errorf("invalid priority: %s", *req.Priority)
	}

	return nil
}

// validateFilter validates a todo filter
func (s *TodoService) validateFilter(filter *models.TodoFilter) error {
	if filter.Priority != nil && !models.IsValidPriority(*filter.Priority) {
		return fmt.Errorf("invalid priority filter: %s", *filter.Priority)
	}

	if filter.SortBy != "" {
		validSortFields := map[string]bool{
			"createdAt": true,
			"title":     true,
			"priority":  true,
			"dueDate":   true,
		}
		if !validSortFields[filter.SortBy] {
			return fmt.Errorf("invalid sort field: %s", filter.SortBy)
		}
	}

	if filter.Order != "" && filter.Order != "asc" && filter.Order != "desc" {
		return fmt.Errorf("invalid sort order: %s (must be 'asc' or 'desc')", filter.Order)
	}

	return nil
}

File: todo-app/frontend/src/services/todoService.js
import axios, { AxiosResponse } from 'axios';
import {
  Todo,
  CreateTodoRequest,
  UpdateTodoRequest,
  TodoFilter,
  APIResponse,
  TodoListResponse,
  TodoStatsResponse,
} from '../types';

// Create axios instance with base configuration
const api = axios.create({
  baseURL: '/api',
  timeout: 10000,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Response interceptor to handle API responses
api.interceptors.response.use(
  (response: AxiosResponse<APIResponse>) => {
    // If the response is successful, return the data
    if (response.data.success) {
      return response;
    }
    // If the API returns success: false, throw an error
    throw new Error(response.data.error || 'API request failed');
  },
  (error) => {
    // Handle network errors or HTTP error status codes
    const message = error.response?.data?.error || error.message || 'Network error';
    throw new Error(message);
  }
);

// Todo API service class
export class TodoService {
  // Get all todos with optional filtering
  static async getAllTodos(filter?: TodoFilter): Promise<TodoListResponse> {
    const params = new URLSearchParams();
    
    if (filter) {
      if (filter.completed !== undefined) {
        params.append('completed', filter.completed.toString());
      }
      if (filter.priority) {
        params.append('priority', filter.priority);
      }
      if (filter.tag) {
        params.append('tag', filter.tag);
      }
      if (filter.sortBy) {
        params.append('sortBy', filter.sortBy);
      }
      if (filter.order) {
        params.append('order', filter.order);
      }
    }
    
    const queryString = params.toString();
    const url = queryString ? `/todos?${queryString}` : '/todos';
    
    const response = await api.get<APIResponse<TodoListResponse>>(url);
    return response.data.data!;
  }

  // Get a specific todo by ID
  static async getTodoById(id: string): Promise<Todo> {
    const response = await api.get<APIResponse<Todo>>(`/todos/${id}`);
    return response.data.data!;
  }

  // Create a new todo
  static async createTodo(todo: CreateTodoRequest): Promise<Todo> {
    const response = await api.post<APIResponse<Todo>>('/todos', todo);
    return response.data.data!;
  }

  // Update an existing todo
  static async updateTodo(id: string, updates: UpdateTodoRequest): Promise<Todo> {
    const response = await api.put<APIResponse<Todo>>(`/todos/${id}`, updates);
    return response.data.data!;
  }

  // Delete a todo
  static async deleteTodo(id: string): Promise<void> {
    await api.delete(`/todos/${id}`);
  }

  // Toggle todo completion status
  static async toggleTodo(id: string): Promise<Todo> {
    const response = await api.patch<APIResponse<Todo>>(`/todos/${id}/toggle`);
    return response.data.data!;
  }

  // Get todo statistics
  static async getStats(): Promise<TodoStatsResponse> {
    const response = await api.get<APIResponse<TodoStatsResponse>>('/stats');
    return response.data.data!;
  }

  // Health check
  static async healthCheck(): Promise<boolean> {
    try {
      await api.get('/health');
      return true;
    } catch (error) {
      return false;
    }
  }
}

// Export utility functions for error handling
export const handleApiError = (error: any): string => {
  if (error.response?.data?.error) {
    return error.response.data.error;
  }
  if (error.message) {
    return error.message;
  }
  return 'An unexpected error occurred';
};

// Export the axios instance for custom requests if needed
export { api };

// Default export
export default TodoService;